plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/XProjectManager.java

```diff
@@ -113,23 +125,30 @@ public void initialize(ProjectDescription description, IProjectConfig projectCon
 	/** Initial build reads the project state and resolves changes. */
 	public XBuildResult doInitialBuild(CancelIndicator cancelIndicator) {
 		Set<URI> changedSources = projectStateHolder.readProjectState(projectConfig);
-
+		// TODO distinguish between changed sources and deleted sources
 		XBuildResult result = doIncrementalBuild(changedSources, Collections.emptySet(),
 				Collections.emptyList(), cancelIndicator);
 
-		if (!changedSources.isEmpty()) {
-			projectStateHolder.writeProjectState(projectConfig);
+		// clear the resource set to release memory
+		boolean wasDeliver = resourceSet.eDeliver();
+		try {
+			resourceSet.eSetDeliver(false);
+			resourceSet.getResources().clear();
+		} finally {
+			resourceSet.eSetDeliver(wasDeliver);
 		}
+
+		persistProjectState();
```

Comment:
~Did you make sure to only do this if persisting wasn't turned off via command line option?~
(all good; it's being checked in the ProjectStateHolder)